src/main/scala/cromwell/binding/Workflow.scala

```diff
@@ -2,14 +2,89 @@ package cromwell.binding
 
 import cromwell.binding.AstTools.{AstNodeName, EnhancedAstNode}
 import cromwell.binding.types.WdlType
-import cromwell.parser.WdlParser.{Ast, SyntaxError, Terminal}
+import cromwell.parser.WdlParser.{Ast, AstList, SyntaxError, Terminal}
 
+import scala.collection.JavaConverters._
+import scala.collection.mutable
 import scala.language.postfixOps
 
-
 object Workflow {
 
-  def apply(ast: Ast, wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter, parent: Option[Scope]): Workflow = {
+  def apply(ast: Ast,
+            namespaces: Seq[WdlNamespace],
+            tasks: Seq[Task],
+            wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter): Workflow = {
+    ast.getName match {
+      case AstNodeName.Workflow =>
+        generateWorkflowScopes(ast, namespaces, tasks, wdlSyntaxErrorFormatter)
+      case nonWorkflowAst =>
+        throw new UnsupportedOperationException(s"Ast is not a 'Workflow Ast' but a '$nonWorkflowAst Ast'")
+    }
+  }
+
+  private def generateWorkflowScopes(workflowAst: Ast,
+                                     namespaces: Seq[WdlNamespace],
+                                     tasks: Seq[Task],
+                                     wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter): Workflow = {
+
+    /**
+     * Incremented indexes per type of class. Currently only used by Scatter.
+     * Indexes are incremented just before use + recursion (prefix, vs. infix or postfix),
+     * thus "start" at zero, but default to -1.
+     */
+    val scopeIndexes: mutable.Map[Class[_ <: Scope], Int] = mutable.HashMap.empty.withDefaultValue(-1)
+
+    /**
+     * Retrieve the list of children Asts from the AST body.
+     * Return empty seq if body is null or is not a list.
+     */
+    def getChildrenList(ast: Ast): Seq[Ast] = Option(ast.getAttribute("body")) collect {
```

Comment:
This looks like the same pattern I complained about elsewhere. The collect getOrElse something which is then collected on all in a single expression spread across multiple lines is hard to read